### DISCUSSION TITLE
Tasmota script - sendmail not working reliably (13.4.0 on ESP32 DevKitC4) ?

### DISCUSSION THREAD

**Discussion by forumlw06** [arendst/Tasmota#20948] [Q&A] [answered]
Hello,

i have a ATHOM PLUG V3 (ESP32C3) setup to work as a central "mail send device" (other devices use that device to send mails on events, so that i do only need one device with sendmail enabled in my network). This works very reliably.

I wanted to free up this ATHOM plug and replace it by an ESP32 DevKitC V4 board (4MB flash, ESP Chip ID	7606372 (ESP32-D0WD-V3 v3.0)).

I used the same defines in Tasmocompiler as on my ATHOM PLUG, but somehow the DevKit V4 board behaves different.

When i run this script:
```
>D 48
IP=192.168.178.61

sm=1

>S
; every 30s
if (secs%30==0) {
	if (sm==1) {
		print --- start email heap=%heap%
		=>sendmail [*:*:*:*:*:tomail@web.de:ESP32 MTEST] Bodytest
		print --- stop  email heap=%heap%
		print ---
	} else {
		print --- sm disabled
	}
}
```

i get this output: some mails send successfully, then one with "wrong parameter" even though the mail parameters were hard coded via defines. Also the device seems to get very unresponsive (switching to from for example console to main menu takes very long.
May it be a memory leak?? Not sure how to debug other than printing the heap information 
```
22:27:32.549 SCR: nv=1, tv=0, vns=3, vmem=32, smem=16384
22:28:00.905 --- start email heap=147624.00
22:28:00.907 SCR: performs "sendmail [*:*:*:*:*:tomail@web.de:ESP32 MTEST] Bodytest"
22:28:00.909 MAI: Size 41
22:28:00.910 MAI: smtp.gmail.com, 465, xxx@gmail.com, aaaa bbbb cccc dddd
22:28:00.922 MAI: xxx@gmail.com, tomail@web.de, ESP32 MTEST,  Bodytest
22:28:37.282 RSL: RESULT = {"Sendmail":"Done"}
22:28:37.285 --- stop  email heap=145304.00
22:28:37.286 ---
22:29:00.421 --- start email heap=147524.00
22:29:00.422 SCR: performs "sendmail [*:*:*:*:*:tomail@web.de:ESP32 MTEST] Bodytest"
22:29:00.424 MAI: Size 41
22:29:00.425 MAI: smtp.gmail.com, 465, xxx@gmail.com, aaaa bbbb cccc dddd
22:29:00.437 MAI: xxx@gmail.com, tomail@web.de, ESP32 MTEST,  Bodytest
22:29:26.571 RSL: RESULT = {"Sendmail":"Done"}
22:29:26.574 --- stop  email heap=145412.00
22:29:26.575 ---
22:29:30.699 --- start email heap=147828.00
22:29:30.701 SCR: performs "sendmail [*:*:*:*:*:tomail@web.de:ESP32 MTEST] Bodytest"
22:29:30.703 MAI: Size 41
22:29:30.704 MAI: smtp.gmail.com, 465, xxx@gmail.com, aaaa bbbb cccc dddd
22:29:30.716 MAI: xxx@gmail.com, tomail@web.de, ESP32 MTEST,  Bodytest
22:29:44.830 RSL: RESULT = {"Sendmail":"Done"}
22:29:44.834 --- stop  email heap=144064.00
22:29:44.835 ---


22:30:01.000 --- start email heap=147412.00
22:30:01.002 SCR: performs "sendmail [*:*:*:*:*:tomail@web.de:ESP32 MTEST] Bodytest"
22:30:01.004 MAI: Size 41
22:30:01.005 MAI: smtp.gmail.com, 465, xxx@gmail.com, aaaa bbbb cccc dddd
22:30:01.017 MAI: xxx@gmail.com, tomail@web.de, ESP32 MTEST,  Bodytest
22:30:45.717 RSL: RESULT = {"Sendmail":"Wrong parameters"}
22:30:45.721 --- stop  email heap=143788.00
22:30:45.722 ---


22:31:00.831 --- start email heap=145668.00
22:31:00.832 SCR: performs "sendmail [*:*:*:*:*:tomail@web.de:ESP32 MTEST] Bodytest"
22:31:00.835 MAI: Size 41
22:31:00.836 MAI: smtp.gmail.com, 465, xxx@gmail.com, aaaa bbbb cccc dddd
22:31:00.848 MAI: xxx@gmail.com, tomail@web.de, ESP32 MTEST,  Bodytest
22:31:54.308 RSL: RESULT = {"Sendmail":"Done"}
22:31:54.311 --- stop  email heap=144380.00
22:31:54.313 ---
22:32:00.371 --- start email heap=148468.00
22:32:00.372 SCR: performs "sendmail [*:*:*:*:*:tomail@web.de:ESP32 MTEST] Bodytest"
22:32:00.375 MAI: Size 41
22:32:00.376 MAI: smtp.gmail.com, 465, xxx@gmail.com, aaaa bbbb cccc dddd
22:32:00.388 MAI: xxx@gmail.com, tomail@web.de, ESP32 MTEST,  Bodytest
22:32:34.682 RSL: RESULT = {"Sendmail":"Done"}
22:32:34.686 --- stop  email heap=145296.00
```

As Gerhard (gemu2015) suggested in one of my earlier questions, i increased the stack size. Here my (mail related) defines:
```
//-------------------------------
//-- sendmail additions
//-------------------------------
#ifndef USE_SENDMAIL
	#define USE_SENDMAIL
#endif

#define DEBUG_EMAIL_PORT // this is the magic bullet for all those {"Sendmail":"Wrong parameters"}
//#define MAIL_TIMEOUT 2000

#define EMAIL_SERVER "smtp.gmail.com"
#define EMAIL_PORT 465
#define EMAIL_USER "xxx@gmail.com"
#define EMAIL_PASSWORD "aaaa bbbb cccc dddd" // generated by gmail as application specific password

#ifdef ESP32
	//-- for 13.2.0 and above the new sendmail library needs more stack size on esp32 => increase stacksize 
	#undef SET_ESP32_STACK_SIZE
	#define SET_ESP32_STACK_SIZE (12 * 1024)
	//-- FROM without <> on ESP32
	#define EMAIL_FROM "xxx@gmail.com"
#endif
```

Thanks!

Regards Rainer

**Comment by Jason2866** [comment]
Try with Arduino Core 3.0.0. We are in process of migrating too. There is no investigation done related/caused to core 2.0.14 issues.

**Reply by gemu2015** [reply]
Compile with DEBUG_EMAIL_PORT (comment out) in Webserver email32
Will show details why email fails

**Reply by gemu2015** [reply]
btw. be warned.  you may not send emails too often:
when i developed this driver i did send a lot of emails for testing.
suddenly my account was blocked for at least a week. i was considered as a spammer.
until that day some of my emails to friends go the spam folder. it seems i am on a blacklist

**Reply by forumlw06** [reply]
Thanks Jason/Gerhard.

As i am a happy user of Tasmota (not a developer) Jason's suggestion sounds "heavy"/undoable for me (or is changing to Arduino Core 3.0.0 somewhere documented?). Probably have to wait until this is standard base for Tasomota? Is there a timeline, when Tasmota si planned to be on that Arduino Core?

@Gerhard: DEBUG_EMAIL_PORT was enabled in my build (see above), but probably i also should have set Weblog to 4 to see more details? 
I will try to catch more details with increased weblog, and post if i am able to catch one.
What really puzzles me is, that my ESP32C3 build (same build flags) on my ATHOM Plug V3 runs very reliably. Also i did not see any "slow downs" when switching between menus etc. Whereas on my ESP32 DevKit 4 Board sendmail runs for some time, then suddenly i get "wrong parameters" and then it switches back to normal sending. Also that slow reaction of the board is different from my ESP32C3 build.

And yes: i am aware of being potentially marked as a spammer, that's why i don't let run my tests too long (and have used a new account in case i am banned). Thanks for the headsup

**Comment by ghost** [comment]
I also have an ESP32C3 and use the following setting in the "user_config_override.h" file:
#define USE_ESP32MAIL
#define USE_TLS
#define EMAIL_USER "***"
#define EMAIL_PASSWORD "***"
#define EMAIL_FROM "***"
#define EMAIL_SERVER "***"
#define EMAIL_PORT 465
#define MAIL_TIMEOUT 2000

in the "script":
{
=>sendmail [*:*:*:*:*:EMail-Adress-to-send:reference]*
>m
dstart: %dstart% - mstart: %mstart% - pstart: %pstart% - dstat2: %dstat2%
&sday
&dcon
&mcon
&pcon
&dpro
#
}

**Reply by forumlw06** [reply]
Thanks for the information. I do have sendmail running very reliably on my ATHOM Plug V3 (ESP32C3) with the defines you mention above. But with the same defines i do get "wrong parameters" during mail send from time to time on an ESP32 Devkit4 Board (not always and not reproducable) and the device gets really unresponsive then. Unfortunately i was not able to solve that until now ... I wanted to move my sendmail to a ESP32 Devkit Board to also connect some sensors (DS18B20, DHT22, ...)

**Reply by gemu2015** [reply]
detailed debug was not in function with current source.
attached is an update to the driver with better debug. replace the file in tasmota build
and enable debug in user config with
#define DEBUG_EMAIL_PORT


[xdrv_01_2_webserver_esp32_mail.ino.zip](https://github.com/arendst/Tasmota/files/14662871/xdrv_01_2_webserver_esp32_mail.ino.zip)

**Reply by forumlw06** [reply]
Thank you Gerhard! First had to find out how to replace that file in tasmocompiler, but i hope i was successful in doing so. Unfortunately i still do not get more information that before, even though i have DEBUG_EMAIL_PORT active.

My current script looks as follows (trying to send mail every 30 seconds
```>D 48
sm=1

>S
if (secs%30==0) {
	if (sm==1) {
		print --- start email heap=%heap% # stack=%stack% # ram=%ram%
		=>sendmail [*:*:from@gmail.com:aaaa bbbb cccc dddd:from@gmail.com:to@web.de:ESP32 MTEST] Bodytest
		print --- stop  email heap=%heap% # stack=%stack% # ram=%ram%
		print ---
	} else {
		print --- sm disabled
	}
}
```

extract from my log (weblog 4) looks like this:
```
00:32:30.985 CMD: Grp 0, Cmd 'SENDMAIL', Idx 1, Len 106, Pld -99, Data '[*:*:from@gmail.com:aaaa bbbb cccc dddd:from@gmail.com:to@web.de:ESP32 MTEST] Bodytest'
00:32:30.988 MAI: Size 95
00:32:30.989 MAI: smtp.gmail.com, 465, from@gmail.com, aaaa bbbb cccc dddd
00:32:30.991 MAI: from@gmail.com, to@web.de, ESP32 MTEST,  Bodytest
00:33:01.556 RSL: RESULT = {"Sendmail":"Wrong parameters"}
00:33:01.560 --- stop  email heap=136304.00 # stack=9680.00 # ram=16466.00
00:33:01.572 ---
00:33:20.618 WIF: Prüfe Verbindung…
00:33:30.592 --- start email heap=142104.00 # stack=9680.00 # ram=16466.00
00:33:30.594 SCR: performs "sendmail [*:*:from@gmail.com:aaaa bbbb cccc dddd:from@gmail.com:to@web.de:ESP32 MTEST] Bodytest"
00:33:30.606 SRC: Rule
00:33:30.608 CMD: Grp 0, Cmd 'SENDMAIL', Idx 1, Len 106, Pld -99, Data '[*:*:from@gmail.com:aaaa bbbb cccc dddd:from@gmail.com:to@web.de:ESP32 MTEST] Bodytest'
00:33:30.609 MAI: Size 95
00:33:30.611 MAI: smtp.gmail.com, 465, from@gmail.com, aaaa bbbb cccc dddd
00:33:30.612 MAI: from@gmail.com, to@web.de, ESP32 MTEST,  Bodytest
00:34:01.180 RSL: RESULT = {"Sendmail":"Wrong parameters"}
00:34:01.184 --- stop  email heap=137472.00 # stack=9680.00 # ram=16466.00
00:34:01.196 ---
00:34:10.972 WIF: Prüfe Verbindung…
```

You can see that it takes 31 seconds from "--- start email" to "--- stop email" debug information that i added. This somehow corresponds to the very bad response of the device if i try to switch from console to e.g. main menu of tasmota. 
Device seems to be very busy with something, i just do not have any clue with what ...

Not even sure, why i do not have a single successful mail send at the moment (yesterday many sends were successful, and only from time to time i got "wrong parameter". Really strange!)

My mail credentials as well as the app password work reliably on my ESP32C3 plug, so i would assume that this is principially working fine. (BTW one difference between plug and Devkit Board is that on the plug i added the required energy sensors whereas on the board i added DS18B20 (+ Alias), but i would assume that this would not matter.

Anyhow: i am by far not as deep into the topic as you. Maybe i am doing something wrong during compile with tasmocompiler (although i have been using that for a while adding script + google chart functionality to my various plugs. I really like that and these "special builds" work).

Regards 

Rainer

**Reply by forumlw06** [reply]
BTW, i was replacing your file in 13.4.0

**Reply by gemu2015** [reply]
i don't think you can easily exchange source files on tasmocompiler
and yes sending emails takes some time, 30 seconds is not uncommon.
if you want to do that in the background, enable multitasking
#define USE_SCRIPT_TASK

you will see that tasmota is no longer blocked during email sent

```
>D

>B
; create task 1 every 60 seconds on core1, priority 3, stack 12kb
ct(1 60000 1 3 12000)

; task 1
>t1
print sending email
=>sendmail [*:*:*:*:*:mrx@gmail.com:test] Hallo
print email sent

>S

>W
Seconds{m}%upsecs%
```

**Reply by forumlw06** [reply]
Thanks, actually i think i managed to exchange the file on tasmocompiler, but i need to verify, this really was used. Will play a bit during the weekend and probably start working into VSCode/Platform IO locally. On the long run that seems to be better. Also will play with the tasking suggestion above. Many things to explore/learn/fail ;-). Steep and stony learning curve ...

Would the code above send a mail every 60 secs? Would it be possible to run sendmail in a seperate task not every 60 secs, but only when special conditions (like a specific temperature, ... ) is met?

**Reply by gemu2015** [reply]
on demand example

```
>D
sm=0

>B

; create task 1 every  second on core1, priority 3, stack 12kb
ct(1 1000 1 3 12000)

>t1
if sm>0 {
print sending email
=>sendmail [*:*:*:*:*:mrx@gmail.com:test] Hallo
print email sent
sm=0
}

>S

>W
; send email on button
bu(sm "pending mail" "Send mail") 
Seconds{m}%upsecs%
```

**Reply by forumlw06** [reply]
Hello Gerhard,

thank you for the tips. I tried your suggestion with the tasks, but sendmail also did not work in that configuration ("wrong paramteres" and Tasmota was kind of frozen).

As i have been playing too long with that without success, i decided to step back and restart to see if i can find the difference between my working ESP32C3 plug
and the non working ESP32 Devkit V4 board.

So i first compiled Tasmota 13.4.0 (Tasmocompiler) for the board with the same settings as for the plug (even though the energy sensors of the pug ar not available on the board).
Tasmota was very responsive and mails are sent within 2-4 seconds.

I then added DS18B20 (plus DS18Alias), this also works. 

**But as soon as i add for example DHT (#define USE_DHT) or full webpage (#define SCRIPT_FULL_WEBPAGE)
i see the extreme slow down of Tasmota (nearly unresponsive) and sendmail returns "wrong paramters" (after 40-50 seconds).**

So i assume, there is not enough resouces (ram, stack??) and maybe that also was the case, when i tried your task example above (in that build, i had DHT active in addition to DS18B20/DS18Alias).

Interestingly if i monitor heap/stack/ram (print every 5 seconds to console in my script), i see a big difference between my ESPC3 ATHOM plug and the ESP32 Devkit V4 Board.

**ESP32C3 has nearly 3 times more stack space and 1,5 times more heap.**

Might that be the problem? And why is there so a big difference between the two (when using the same tasmota features, ok, RISC vs Tencilla architecture, but not sure if that really explains the difference). By looking at the specs, i would have thought that they are similar

_**ESP32 Devkit V4 Board:**_

- heap=135352.00 # stack=9680.00 # ram=16466.00

**_ESP32C3 ATHOM PLUG_**

- heap=181260.00 # stack=31904.00 # ram=16466.00

Too bad that i do not get the board with the DS18B20 and DHT sensors working reliably with sendmail. That would be my preferred solution, but maybe i have to use 2 devices to achive, what i wanted.

**Comment by forumlw06** [comment]
I just managed to compile 13.4.0 with the "xdrv_01_2_webserver_esp32_mail.ino" you provided above and all required drivers (DS18B20/DHT/FULL_WEB_PAGE) directly on Gitpod (not susing Tasmocompiler). To be sure, that i used the driver provided by you, i replaced "MAI:" with "M_I" so that i see on console, that i have been using your file. 

**And guess what: no sending mail works fast (2-3 seconds), tasmota is responsive and i get this as log. Not sure, why this works with your driver??**
Here is an extract of the log:
```
18:04:55.705 -- 55.00 -- heap=134680.00 # stack=9680.00 # ram=16466.00
18:04:57.724 DHT: Pin13 cycles (80/80)  165 68 164 80 165 79 165 80 165 79 165 79 165 80 165 223 206 226 165 80 ..
18:04:57.726 DHT: Pin13 read 01A300EC90
18:05:00.703 --- start email heap=134680.00 # stack=9680.00 # ram=16466.00
18:05:00.706 SCR: performs "sendmail [*:*:from@gmail.com:aaaa bbbb cccc dddd:from@gmail.com:to@web.de:ESP32 MTEST] Bodytest"
18:05:00.720 SRC: Rule
18:05:00.721 CMD: Grp 0, Cmd 'SENDMAIL', Idx 1, Len 106, Pld -99, Data '[*:*:from@gmail.com:aaaa bbbb cccc dddd:from@gmail.com:to@web.de:ESP32 MTEST] Bodytest'
18:05:00.725 M_I Size 95
18:05:00.727 M_I smtp.gmail.com, 465, from@gmail.com, aaaa bbbb cccc dddd
18:05:00.728 M_I from@gmail.com, to@web.de, ESP32 MTEST,  Bodytest
18:05:00.743 #### Connecting to SMTP server...
18:05:01.308 
#### SMTP server connected
18:05:01.314 
#### Sending greeting response...
18:05:01.375 
#### Logging in...
18:05:01.708 
#### Sending Email...
18:05:01.710 
#### Sending message header...
18:05:01.782 
#### Sending message body...
18:05:02.021 
#### Finishing the message sending...
18:05:02.730 
#### Closing the session...
18:05:02.779 
#### Message sent successfully
18:05:02.791 
18:05:02.792 Message sent success: 1
18:05:02.794 Message sent failled: 0
18:05:02.796 Message No: 1
18:05:02.797 Status: success
18:05:02.799 Date/Time: 2024/3/24 18:5:2
18:05:02.810 Recipient: <��?
18:05:02.811 Subject: ��?
18:05:02.817 RSL: RESULT = {"Sendmail":"Done"}
18:05:02.821 --- stop  email heap=133588.00 # stack=9680.00 # ram=16466.00
```

Did you change more than just re-enabling the debug information?
Why are Recipient/Subject looking so strange (log above)?

Anyhow, maybe that's the solution. Will build a version with your driver but without mail debug enabled.

**Reply by gemu2015** [reply, marked as answer]
i did not change anything but the debug feature. meanwhile these changes are merged into Tasmota.
also the bug with Recipient/Subject is fixed.

the esp32 family has different amounts of SRAM available per variant.
i generally prefer the S3 which is by far the most powerful  chip

some sensors may be a problem because they disable Interrupts or multitasking (like DHT or DS18B20)

DHT is not recommended at all.  i prefer Sensirion SHT3x or Bosch BMx chips.

**Reply by Jason2866** [reply]
Probably doing a build with TasmoCompiler adds additional stuff which makes things go to the south.

**Reply by forumlw06** [reply]
Thank you for your help. I just got an S3 Board (16MB flash, 8MB PSRAM (i know i only could use 4MB)), maybe i switch to that (have to solder the pins first). 

Thanks for the info on the sensors. DS18B20 would be needed, as i can put them into liquids (like own brewed beer :-) ). As DHT was lying around i thought i use it to monitor room temperature as well.

Any idea, why my 13.4.0 tasmota32.bin from gitpod is 1,93 MB, while the tasmocompiler one is only 1,30 MB (same user_config_override.h)?

I was just wondering to see the big difference.

**Reply by Jason2866** [reply]
S3 can use 8MB PSRAM. The 4MB PSRAM limit is only ESP32. TasmoCompiler is not maintained from Tasmota dev crew. So we can't tell anything how it manage the builds.

**Reply by forumlw06** [reply]
Ah, thanks for the info on S3 PSRAM. 

How do i need to compile it in Gitpod to use that? Sorry for my newbee questions. Gitpod compiles are very new for me ...

**Reply by Jason2866** [reply]
There are different S3 variants regarding flash and PSRAM type. First you have to find out which type of flash and PSRAM your device has. QIO or OPI?
When you know this, you just take the corresponding S3 boards name in your env.
Example https://github.com/arendst/Tasmota/blob/15ae3e0fa5d6790925e9e50aa9d78b840d7cb8ed/platformio_tasmota_cenv_sample.ini#L36-L46

**Reply by forumlw06** [reply]
I see! Not sure, what type of flash and PSRAM i have. Didn't even know that there are differences :-( (newbee). 

I barely dare to ask, how i can find that information (i did not to any research so far), but maybe the answer is simple?. If not, just forget about it. Need to do some research and then come back otherwise.

**Reply by Jason2866** [reply]
There is a laser print on the metal case. Something like N16R8. Google for the S3 type number your module has.

**Reply by forumlw06** [reply]
Thanks Jason, i do have exactly ESP32-S3-WROOM-1 N16R8. That means i do have 16 MB Quad-SPI Flash and 8 MB Octal-SPI SRAM.

I was building a development tasmota with "platformio run -e tasmota32s3-qio_opi-all".
On the information page in main menu i see the 8 MB PSRAM and the 16 MB flash! Great!

As i do not use WEBCAM/LVGL i guess i can remove the according build flags in [env:tasmota32s3-qio_opi-all]?

As said, i need to dig into that more. Thanks for all your help so far!

Have a nice evening!
